### PR TITLE
KYC Sanction  Expiration

### PR DESCRIPTION
**Pull request by ylv-io**
# Description

This pull request introduces several enhancements and modifications to the KintoID contract, focusing on governance, sanction management, and role definitions. It also updates the IKintoID interface to reflect these changes.

Implementation Details:
- Added a new constant `GOVERNANCE_ROLE` to define a governance role within the contract.
- Introduced two new immutable addresses: `walletFactory` and `faucet`, which are now public and override existing definitions.
- Deprecated the `walletFactory` address by making it private and adding a notice for its deprecation.
- Added a new mapping `sanctionedAt` to track the timestamp of the last sanction applied to each user.
- Updated the `revokeKYC` function to set the `sanctionedAt` timestamp to the current block timestamp when a KYC is revoked.
- Modified the `addSanction` function to update the `sanctionedAt` timestamp when a sanction is added.
- Enhanced the `isKYC` function to include logic for checking if an account is sanctions safe and handling the 3-day confirmation period for sanctions.
- Updated the `isSanctionsSafe` function to incorporate the 3-day confirmation logic for sanctions and to check if an account is monitored for sanctions.
- Introduced a new function `confirmSanction` that allows the governance role to reset the sanction timestamp, making the sanction indefinite.
- Corrected a typo in the `isIndividual` function's return description from "indivdual" to "individual".
- Updated the `IKintoID` interface to include the new `GOVERNANCE_ROLE` function.

## Type of change

- [ ] Bug fix 
- [x] New feature 
- [ ] Breaking change 
- [ ] Dependency changes
- [ ] Deployment
- [ ] Forge Script
- [ ] Code refactor / cleanup
- [ ] Documentation or wording changes
- [ ] Other

## Checklist:

- [x] The diff is legible and has no extraneous changes
- [x] Complex code has been commented, including external interfaces
- [ ] Tests have 100% code coverage
- [x] The base branch is either `main`, or there's a description of how to merge

### CODE REVIEW

File: src/KintoID.sol [hunk 1]
```diff
@@ -219,6 +227,9 @@ contract KintoID is
         Metadata storage meta = _kycmetas[_signatureData.signer];
         meta.mintedAt = 0;
         meta.updatedAt = 0;
+
+        // Set sanctionedAt to the current block timestamp
```
Comment (rrecuero) [inline src/KintoID.sol:231]: is this needed? When we burn we dont care about this anymore

File: src/KintoID.sol [hunk 2]
```diff
@@ -343,7 +357,24 @@ contract KintoID is
      * @return true if the account has KYC token.
      */
     function isKYC(address _account) external view override returns (bool) {
-        return balanceOf(_account) > 0 && isSanctionsSafe(_account);
+        if (balanceOf(_account) == 0) {
+            // If the account doesn't have a KYC token
```
Comment (rrecuero) [inline src/KintoID.sol:361]: this comment is not correct

Comment (rrecuero) [reply src/KintoID.sol:361]: I believe we should just return false in this case, if there is no NFT it should be false

File: src/KintoID.sol [hunk 3]
```diff
@@ -361,7 +392,25 @@ contract KintoID is
      * @return true if the account is sanctions safe.
      */
     function isSanctionsSafe(address _account) public view virtual override returns (bool) {
-        return isSanctionsMonitored(7) && _kycmetas[_account].sanctionsCount == 0;
+        Metadata storage meta = _kycmetas[_account];
+        if (meta.sanctionsCount > 0 || !isSanctionsMonitored(7)) {
+            // If the account has sanctions
+            if (block.timestamp - sanctionedAt[_account] < 3 days) {
```
Comment (rrecuero) [inline src/KintoID.sol:398]: replace all these lines to

```
return block.timestamp - sanctionedAt[_account] < 3 days
```

File: src/KintoID.sol (hunk 1 above)
Comment (ylv-io) [reply src/KintoID.sol:231]: Yes. Otherwise how we'll unblock a user in 3 days, if the burn is not confirmed?

File: src/KintoID.sol (hunk 2 above)
Comment (ylv-io) [reply src/KintoID.sol:361]: But then KYC providers can just burn KYC and users are locked forever, no?

Comment (ylv-io) [reply src/KintoID.sol:361]: Fixed.

File: src/KintoID.sol [hunk 4]
```diff
@@ -361,7 +372,14 @@ contract KintoID is
...
     function isSanctionsSafe(address _account) public view virtual override returns (bool) {
-        return isSanctionsMonitored(7) && _kycmetas[_account].sanctionsCount == 0;
+        Metadata storage meta = _kycmetas[_account];
+        if (meta.sanctionsCount > 0 || !isSanctionsMonitored(7)) {
+            // If the sanction is not confirmed within 3 days, consider the account sanctions safe
+            return (block.timestamp - sanctionedAt[_account]) > 3 days;
+        }
+
+        // If the account has no sanctions, consider it sanctions safe
+        return true;
```
Comment (rrecuero) [inline src/KintoID.sol:382]: this part needs to check !isSanctionsMonitored(7) as well

File: src/KintoID.sol [hunk 5]
```diff
@@ -371,7 +389,18 @@ contract KintoID is
...
      */
     function isSanctionsSafeIn(address _account, uint16 _countryId) external view virtual override returns (bool) {
-        return isSanctionsMonitored(7) && !_kycmetas[_account].sanctions.get(_countryId);
+        if (_kycmetas[_account].sanctions.get(_countryId) || !isSanctionsMonitored(7)) {
+            // If the sanction is not confirmed within 3 days, consider the account sanctions safe
+            return (block.timestamp - sanctionedAt[_account]) > 3 days;
+        }
+
+        // If the account has no sanctions, consider it sanctions safe
+        return true;
```
Comment (rrecuero) [inline src/KintoID.sol:398]: same, needs still check the monitoring

Comment (rrecuero) [reply src/KintoID.sol:398]: The whole function should be one line with a return with all the conditions

File: src/KintoID.sol (hunk 4 above)
Comment (rrecuero) [reply src/KintoID.sol:382]: same here, recommend just one return, no ifs

**Review by rrecuero — APPROVED** [review]

**Pull request merged**